crates/client/rpc/src/lib.rs

```diff
@@ -866,6 +867,109 @@ where
             None => Err(StarknetRpcApiError::TxnHashNotFound.into()),
         }
     }
+
+    fn get_proof(&self, get_proof_input: RpcGetProofInput) -> RpcResult<RpcGetProofOutput> {
+        if get_proof_input.keys.len() > MAX_KEYS {
+            error!(
+                "Too many keys requested! limit: {:?},
+				requested: {:?}",
+                MAX_KEYS,
+                get_proof_input.keys.len() as u32
+            );
+            return Err(StarknetRpcApiError::ProofLimitExceeded.into());
+        }
+
+        let substrate_block_hash =
+            self.substrate_block_hash_from_starknet_block(get_proof_input.block_id).map_err(|e| {
+                error!("'{e}'");
+                StarknetRpcApiError::BlockNotFound
+            })?;
+
+        let block = get_block_by_block_hash(self.client.as_ref(), substrate_block_hash).unwrap_or_default();
+
+        let global_state_root = block.header().global_state_root;
+
+        let mut state_commitments = self
+            .overrides
+            .for_block_hash(self.client.as_ref(), substrate_block_hash)
+            .state_commitments(substrate_block_hash)
+            .ok_or_else(|| {
+                error!("Failed to retrieve state commitments");
+                StarknetRpcApiError::InternalServerError
+            })?;
+
+        let class_commitment: Option<FieldElement> = Some(state_commitments.class_commitment.commit().into());
+
+        // Generate a proof for this contract. If the contract does not exist, this will
+        // be a "non membership" proof.
+        let contract_proof =
+            state_commitments.storage_commitment.get_proof(Felt252Wrapper(get_proof_input.contract_address));
+
+        let contract_state_hash =
+            match state_commitments.storage_commitment.get(Felt252Wrapper(get_proof_input.contract_address)) {
+                Some(contract_state_hash) => contract_state_hash,
+                None => {
+                    // Contract not found: return the proof of non membership that we generated earlier.
+                    return Ok(RpcGetProofOutput {
+                        state_commitment: Some(global_state_root.0),
+                        class_commitment,
+                        contract_proof,
+                        contract_data: None,
+                    });
+                }
+            };
+
+        // At this point we know the contract exists, so errors should never be thrown!
```

Comment:
In theory but it could still happen and would be fine, but since we got some state hash in the tree that means the contract's state has been committed already so it exists

Reply:
Can you slightly change the sentence so it's meaning is less ambivalent?